### ISSUE TITLE
support funciton* generators as a sort of do notation.

### ISSUE THREAD

**Issue by graingert** [cwmyers/monet.js#103]

**Comment by ulfryk** [comment]
@graingert  can you give an example of how would you like it to look like ?

**Comment by graingert** [comment]
```js
const helloWorld = putStr("Hello")
  .flatMap(putStr(" "))
  .flatMap(putStr("world!"))
  .flatMap(putStr("\n"));
```

```js
const helloWorld = moent.Do(function* () {
  yield putStr("Hello");
  yield putStr(" ");
  yield putStr("world!");
  yield putStr("\n");
});
```

you can also get values out too:
```js
const hamSpam = monet.Do(function* () {
  const x = yield reader();
  const y = x.toLowerCase();
  yield putStr(y);
});

**Comment by sevcsik** [comment]
Example from RxJS:
https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/spawn.md

**Comment by graingert** [comment]
it would have to crash for monads with more than 1 element.

**Comment by graingert** [comment]
see also: https://github.com/pelotom/burrido and https://github.com/pelotom/immutagen

**Comment by graingert** [comment]
sadly it's really slow: https://github.com/pelotom/burrido#caveats

**Comment by sevcsik** [comment]
Despite the limitations, I do think it would be useful for a limited set Monads where there's only one value, like Maybe and Either. Those who want to use it with streams despite the performance penalty can use burrido instead.

**Comment by ulfryk** [comment]
@graingert @sevcsik sorry for not being responsive. This seems to be quite an interesting feature, yet there is still a lot of work to make 'monet' stable and equipped with all needed/requested basic features. So probably it'll not appear in the library in some short time. I'm thinking also how to make it work without crashing lib for ES5 limited environments…

Anyway:
1. PRs are welcome
2. Maybe we should add `Just` another lib like `@monet/generators` and put all generator centric features there?

PS
I would love to see working `do` notation in JS/TS :)

**Comment by sevcsik** [comment]
I sketched an implementation for the do notation (unfortunately `do` is a reserved word so I went with `run`). The first one is Monet-specific and the second one is more generic:

https://repl.it/@sevcsik/generic-monad-comprehension

**Comment by sevcsik** [comment]
The library itself should only contain the `run` function which itself doesn't contain a generator function (a generator expression just yields a function which returns an iterator), so it will work fine in ES5. Of course, to use it, you'll need ES6 generator support.

My biggest concern about writing a separate library is about determining the `unit` method which we should use to wrap the result. Now, I worked it around with optionally accepting the object which holds the `unit` function as the first argument; if the argument isn't supplied, the result won't be wrapped.

But if we would add it to every monadic type, like `Maybe.run` and `Either.run`, this wouldn't be an issue as the run function could just use `this.unit` (and the `do` word is not reserved for for object fields, so there's that :))

**Comment by sevcsik** [comment]
> it would have to crash for monads with more than 1 element.
I agree. Do you know any way to detect that in a generic way? I couldn't think of anything... Of course, we could whitelist monads which are known to work if we only support Monet's monads.

**Comment by sevcsik** [comment]
Opened a PR. I'm yet to figure out how to make the `do` function optional in TS, so it doesn't bring in an ES6 requirement.
